internal/exp/metrics/streams/streams.go

```diff
@@ -0,0 +1,88 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package streams // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/streams"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+)
+
+// Sequence of streams that can be iterated upon
+type Seq[T any] func(yield func(identity.Stream, T) bool) bool
+
+type ErrNotExist struct {
+	Ident identity.Stream
+}
+
+func (e ErrNotExist) Error() string {
+	return fmt.Sprintf("no item for stream %s", e.Ident)
+}
+
+func NotExist(err error) bool {
+	var notExist ErrNotExist
+	return errors.As(err, &notExist)
+}
+
+func Missing(id identity.Stream) error {
+	return ErrNotExist{Ident: id}
+}
+
+func Exists(id identity.Stream, ok bool) error {
+	if !ok {
+		return Missing(id)
+	}
+	return nil
+}
+
+// Map defines a collection of items tracked by a stream-id and the operations
+// on it
+type Map[T any] interface {
+	Load(context.Context, identity.Stream) (T, error)
+	Store(context.Context, identity.Stream, T) error
+	Delete(context.Context, identity.Stream)
+	Items() func(yield func(identity.Stream, T) bool) bool
+	Len() int
+}
+
+func EmptyMap[T any]() HashMap[T] {
+	return HashMap[T]{items: make(map[identity.Stream]T)}
+}
+
+var _ Map[any] = HashMap[any]{}
+
+type HashMap[T any] struct {
+	items map[identity.Stream]T
```

Comment:
If we just use a typedef, we can avoid the extra indirection in the methods. IE:

`type HashMap[T any] map[identity.Stream]T`. Then the methods don't have to do a double dereference (one for the struct itself, and one for `items`)